src/ngx_http_lua_upstream_module.c

```diff
@@ -349,6 +353,55 @@ ngx_http_lua_upstream_set_peer_down(lua_State * L)
 }
 
 
+static int
+ngx_http_lua_upstream_get_chosen_peer(lua_State * L)
+{
+    ngx_str_t                           host;
+    ngx_http_upstream_srv_conf_t       *us;
+    ngx_http_request_t                 *r;
+    ngx_http_request_t                 *fake_r;
+    ngx_peer_connection_t              *peer;
+    int                                 rc;
+
+    if (lua_gettop(L) != 1) {
+        return luaL_error(L, "exactly one argument expected");
+    }
+
+    host.data = (u_char *) luaL_checklstring(L, 1, &host.len);
+
+    us = ngx_http_lua_upstream_find_upstream(L, &host);
+    if (us == NULL) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "upstream not found");
+        return 2;
+    }
+
+    r = ngx_http_lua_get_request(L);
+
+    // we create a fake request object, which we will use to fetch the upstream
+    // peer, because we don't want to pollute the current request object
+    fake_r = ngx_pcalloc(r->pool, sizeof(ngx_http_request_t));
```

Comment:
Hmm, a fake request might not be good enough for side effect isolation. And a 3rd-party upstream balancer module might fetch its own module ctx too.